### PR TITLE
Memory containment for `FIELD_LIST` operands on x86

### PR DESCRIPTION
**Pull request by SingleAccretion**
Enables broader containment for `FIELD_LIST` operands on x86.

The primary benefits come from containing loads that long decomposition has left (e. g. `LCL_FLD`s):
```diff
-       mov      ecx, dword ptr [ebp+08H]
-       mov      edx, dword ptr [ebp+0CH]
-       push     edx
-       push     ecx
+       push     dword ptr [ebp+0CH]
+       push     dword ptr [ebp+08H]
```

Additionally, there is some code simplification enabled by the use of `OperandDesc` in two `inst_*`  functions with this change.

Some some good x86(-only) [diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1630324&view=ms.vss-build-web.run-extensions-tab). A few small regressions because of some worse RA choices.

Note: I think a couple stress runs for this change are warranted, the logic here is not straightforward.

### CODE REVIEW

**Comment by BruceForstall** [comment]
Some nice diffs here. Looking through the diffs, it's easy to see where we could do better (not related to this change), e.g.:
```
       mov      dword ptr [ebp-24H], edx
       mov      dword ptr [ebp-20H], eax
       push     dword ptr [ebp-20H]                 // could just use "push eax"
       push     dword ptr [ebp-24H]                 // could just use "push edx"
```

File: src/coreclr/jit/emitxarch.cpp [hunk 1]
```diff
@@ -4362,6 +4362,24 @@ void emitter::emitIns_C(instruction ins, emitAttr attr, CORINFO_FIELD_HANDLE fld
     emitAdjustStackDepthPushPop(ins);
 }
 
+void emitter::emitIns_A(instruction ins, emitAttr attr, GenTreeIndir* indir)
```
Comment (BruceForstall) [inline src/coreclr/jit/emitxarch.cpp:4365]: Please add a function header comment for this new function

**Comment by BruceForstall** [comment]
One of your changes does retyping, so we end up with diffs like:
```
;  V17 tmp11        [V17,T04] (  3, 32   )    bool  ->  [ebp-2CH]   spill-single-def V11.value(offs=0x01) P-INDEP "field V11.value (fldOffset=0x1)"
...
       mov      cl, byte  ptr [ebp-2CH]
       mov      byte  ptr [esp+01H], cl
=>
       mov      ecx, dword ptr [ebp-2CH]
       mov      byte  ptr [esp+01H], cl
```
Why do we want this? It looks like it doesn't hurt, but...?

**Review by BruceForstall — CHANGES_REQUESTED** [review]
A few comments

**Comment by BruceForstall** [comment]
> The premise is that we'd like to use `push` for `varTypeIsSmall` `(non-candidate) LCL_VAR/LCL_FLD` nodes.

So are there cases where previously we would have something like:
```
       sub       esp, 4
       mov      cl, byte  ptr [ebp-2CH]
       mov      byte  ptr [esp], cl
```
and now we have something like:
```
       push     dword ptr [ebp-2CH]
```
? Assuming the target was a 4-byte slot?

> What kind of assert do you have in mind?

If I add `assert(!id->idGCref());` it never hits in the baseline. This indicates to me that this code never expected GC refs, and we need to do something, if this is really a GC ref.

But it also looks like there are two cases:
```
       push     0
UNDONE: record GCref push [cns]
       push     0
       ; gcr arg push 4
```
why does the second zero here get a GC update but the first goes through the code path that doesn't?

**Comment by SingleAccretion** [comment]
> So are there cases where previously we would have something like ... and now we have something like

I don't believe so, the optimization is for keeping parity with the previous implementation that did this effectively implicitly:
```c
switch (fieldNode->OperGet())
{
    case GT_LCL_VAR:
        // Notably this doesn't handle the FIELD<short>(fieldNode<byte>) case correctly,
        // but that was ok because only very specific promotion-produced IR shapes were expected here.
        inst_TT(INS_push, fieldNode, 0, 0, emitActualTypeSize(fieldNode->TypeGet())); // Note the "emitActualTypeSize".
        break;
```
There are basically three nodes we can see on this path due to the aforementioned decomposition pessimization (that I also would like to remove at some point): `LCL_VAR<field type> | CNS_INT<INT> | <anything INT-sized>`, the scenario that was regressing without the optimization is a promoted small field of an address-exposed struct.

> why does the second zero here get a GC update but the first goes through the code path that doesn't?

If we are looking at the same SPMI diff for `NJulianRuleTest`, then the diff is that these are for the same argument, I believe:
```diff
 IN006f: 00018C push     0
 IN0070: 00018E push     0
+UNDONE: record GCref push [cns]
 IN0071: 000190 push     0
+       ; gcr arg push 4
 IN0072: 000192 mov      ecx, gword ptr [ebp-5CH]
        ; gcrRegs +[ecx]
        ; GC ptr vars -{V41 V45 V61 V100}
 IN0073: 000195 call     System.IO.Strategies.BufferedFileStreamStrategy:WriteSpan()
        ; gcrRegs -[ecx]
-       ; gcr arg pop 1
+       ; gcr arg pop 2

 02    F0 BF ...| 017C        reg EAX becoming dead
 0E    BF 8A ...| 0182        reg ECX becoming dead
 F0 BF 8A    ...| 018C        push ptr  1 (iptr)
-F0 49    C8 ...| 0195        reg ECX becoming live
+A6    0D D0 ...| 0192        push ptr  4
+4B    D0 4B ...| 0195        reg ECX becoming live
 0D    4B F2 ...| 019A        reg ECX becoming dead
-C8    F2 45 ...| 019A        pop  1 ptrs
+D0    F2 45 ...| 019A        pop  2 ptrs
 4B    45 F1 ...| 019D        reg ECX becoming live
 F2 45    05 ...| 01BA        reg EAX becoming live
 F1 05    40 ...| 01CF        reg EAX becoming dead
```
> This indicates to me that this code never expected GC refs, and we need to do something, if this is really a GC ref.

My understanding of the GC reporting for all `push/pop`s is that it already happens via the `emitOutputInstr` -> [`emitStackPush`](https://github.com/dotnet/runtime/blob/8b42dffa5f37762caddbe949ebc2fc348cca5121/src/coreclr/jit/emitxarch.cpp#L14591) -> [`emitStackPushLargeStk`](https://github.com/dotnet/runtime/blob/8b42dffa5f37762caddbe949ebc2fc348cca5121/src/coreclr/jit/emit.cpp#L8733) -> [`gcRegPtrAllocDsc`](https://github.com/dotnet/runtime/blob/8b42dffa5f37762caddbe949ebc2fc348cca5121/src/coreclr/jit/emit.cpp#L8841) call chain.

**Comment by SingleAccretion** [comment]
> Comments?

I would like to understand - what's the reason for expecting only `null`s here? I think we can get (and should handle) arbitrary constants, `byref`s especially. E. g. here is a construction that would hit the `val != 0` case (there could be others):

```cs
[MethodImpl(MethodImplOptions.NoInlining)]
private static void Problem(ref int addrSrc)
{
    var addr = MemoryMarshal.CreateSpan(ref addrSrc, 1);
    ref int constAddr = ref *(int*)100;  // Just a placeholder, it could be address of a real (pinned) GC object

    if (Unsafe.AreSame(ref MemoryMarshal.GetReference(addr), ref constAddr))
    {
        CallForSpan(addr);
    }
}

[MethodImpl(MethodImplOptions.NoInlining)]
static void CallForSpan<T>(Span<T> value) { }
```

**Comment by BruceForstall** [comment]
> I would like to understand - what's the reason for expecting only `null`s here?

I was just being conservative -- I didn't expect anything else. How could you get a non-null constant byref? I would think your example would be illegal.

**Comment by SingleAccretion** [comment]
> How could you get a non-null constant byref?

Well, we are sort of defining that for ourselves at this point, but in general the compiler does expect to see non-null `byref`s (VN [handles](https://github.com/dotnet/runtime/blob/a5665daf8a6f4e5bf44a5e98a92438edae65239e/src/coreclr/jit/valuenum.cpp#L7646-L7665) them at the very least least), and we had [bugs](https://github.com/dotnet/runtime/issues/61510) caused by "only `null` byrefs exist"-like assumptions.

It is true that we don't have that many sources of them right now, but it also the case that with the `Unsafe.NullRef` and derivatives I don't think we can really say that it's illegal at this point.

> I would think it would have to be pinned memory if it was in the GC heap, at which point it could just be an unmanaged pointer, but it couldn't be constant anyway.

Ahh, but it could. Say it was a pointer into an array allocated with `GC.AllocateArray(length, pinned: true)`, that was then saved into a `static readonly`, which would in turn be imported as a constant. Or that the creator of the code was actually using Reflection.Emit and "burned-in" the address of the object.

We also have the curious case of collectible assemblies which can be kept alive by RVAs-turned-constant-byrefs pointing into them.

**Comment by BruceForstall** [comment]
Ok, you've convinced me there are weird cases, so we don't need to assert that the constant is null.

But it seems it's still safe to remove the GC reporting on it, since obviously the constant can't be updated by the GC.

**Review by BruceForstall — APPROVED** [review]
Thanks for all the explanations and updates.

**Comment by kunalspathak** [comment]
Another windows/x86 regression in https://github.com/dotnet/perf-autofiling-issues/issues/4380

**Pull request merged**